### PR TITLE
fix heap cost calculation rounding error

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
Should round up heap size to next 32k block before `div`

#### Summary of Changes
- refactor the heap cost into a function
- add test with refraction heap size (the 3rd case would fail before fix)
- fix size truncation by round up to next 32K block first
- add feature gate
- check result of `consume_checked` if feature is activated 

OPTIONAL: Feature Gate Issue: #30679

### CODE REVIEW

**Comment by taozhu-chicago** [comment]
@Lichtso if this is a valid change, it probably needs a feature gate, or perhaps can be rolled into other feature gates if possible?

**Comment by Lichtso** [comment]
Yes, it needs a feature gate, and I can't think of any other related one to merge this change into.

**Comment by taozhu-chicago** [comment]
> Yes, it needs a feature gate, and I can't think of any other related one to merge this change into.

Thanks for quick response. Feature gate added.

File: programs/bpf_loader/src/lib.rs [hunk 1]
```diff
@@ -307,11 +321,13 @@ pub fn create_ebpf_vm<'a, 'b>(
     orig_account_lengths: Vec<usize>,
     invoke_context: &'a mut InvokeContext<'b>,
 ) -> Result<EbpfVm<'a, RequisiteVerifier, InvokeContext<'b>>, EbpfError> {
-    let _ = invoke_context.consume_checked(
-        ((heap.len() as u64).saturating_div(32_u64.saturating_mul(1024)))
-            .saturating_sub(1)
-            .saturating_mul(invoke_context.get_compute_budget().heap_cost),
-    );
+    let _ = invoke_context.consume_checked(calculate_heap_cost(
```
Comment (Lichtso) [inline programs/bpf_loader/src/lib.rs:324]: Now that we are feature gating this anyway, we should also stop discarding / ignoring the result. In other words `let _ =` should become `let heap_cost_result = ...; if feature { heap_cost_result?; }`.

Comment (taozhu-chicago) [reply programs/bpf_loader/src/lib.rs]: Yea sure, added it.

File: programs/bpf_loader/src/lib.rs [hunk 2]
```diff
@@ -298,6 +298,20 @@ fn check_loader_id(id: &Pubkey) -> bool {
         || bpf_loader_upgradeable::check_id(id)
 }
 
+fn calculate_heap_cost(heap_size: u64, heap_cost: u64, enable_rounding_fix: bool) -> u64 {
+    const KILOBYTE: u64 = 1024;
+    const PAGE_SIZE_KB: u64 = 32;
+    let mut rounded_heap_size = heap_size;
+    if enable_rounding_fix {
+        rounded_heap_size = rounded_heap_size
+            .saturating_add(PAGE_SIZE_KB.saturating_sub(1).saturating_mul(KILOBYTE));
```
Comment (Lichtso) [inline programs/bpf_loader/src/lib.rs:307]: Shouldn't it be `PAGE_SIZE_KB.saturating_mul(KILOBYTE).saturating_sub(1)` (order of mul and sub swapped)?

File: programs/bpf_loader/src/lib.rs [hunk 3]
```diff
@@ -298,6 +298,20 @@ fn check_loader_id(id: &Pubkey) -> bool {
...
+    }
+    rounded_heap_size
+        .saturating_div(PAGE_SIZE_KB.saturating_mul(KILOBYTE))
+        .saturating_sub(1)
```
Comment (Lichtso) [inline programs/bpf_loader/src/lib.rs:311]: also, why always subtract one? That way the first 31 KiB don't count.

File: programs/bpf_loader/src/lib.rs [hunk 4]
```diff
@@ -298,6 +298,20 @@ fn check_loader_id(id: &Pubkey) -> bool {
         || bpf_loader_upgradeable::check_id(id)
 }
 
+fn calculate_heap_cost(heap_size: u64, heap_cost: u64, enable_rounding_fix: bool) -> u64 {
+    const KILOBYTE: u64 = 1024;
```
Comment (Lichtso) [inline programs/bpf_loader/src/lib.rs:302]: Nit: This is technically a [kibibyte](https://en.wikipedia.org/wiki/Kibibyte) not a [KB](https://en.wikipedia.org/wiki/Kilobyte). (see https://en.wikipedia.org/wiki/Orders_of_magnitude_(data))

File: programs/bpf_loader/src/lib.rs (hunk 2 above)
Comment (taozhu-chicago) [reply programs/bpf_loader/src/lib.rs:307]: The math should be `(heap_size + (32KB-1))/32KB)`, thanks for catching it.

File: programs/bpf_loader/src/lib.rs (hunk 3 above)
Comment (taozhu-chicago) [reply programs/bpf_loader/src/lib.rs:311]: gotit

File: programs/bpf_loader/src/lib.rs (hunk 4 above)
Comment (taozhu-chicago) [reply programs/bpf_loader/src/lib.rs:302]: TIL kibibyte 👍🏼

File: programs/bpf_loader/src/lib.rs (hunk 3 above)
Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:311]: To keep compatible with the original behavior it should be the following, right?
`.saturating_sub(if enable_rounding_fix { 0 } else { 1 })`

Comment (taozhu-chicago) [reply programs/bpf_loader/src/lib.rs:311]: Mmm, don't think need to do that. The first 31KiB costs zero in both before/after activation. The test at Ln3945 asserts these scenarios.

Comment (Lichtso) [reply programs/bpf_loader/src/lib.rs:311]: If it is intentional then leave it as is.

**Review by Lichtso — APPROVED** [review]

**Pull request merged**